### PR TITLE
PPL command expression implementation for `geoip`

### PR DESCRIPTION
**Pull request by andy-k-improving**
### Description
Introduce a new PPL command expression `geoip`, to perform geo-spatial information lookup with the provided IPv4 || IPv6 addresses, result of the lookup is formatted into a tuple with attribute as key and location detail as value. 

In this particular setting, SQL plugin will act as a thin client, by relaying the IPEnrichment request to OpenSearch Geo-Spatial plugin, WITHIN the same cluster. 
Detail implementation and interface that exposed on Geo-Spatial side can be found:
https://github.com/opensearch-project/geospatial/pull/700

Internally this functionality is achieved by:
 - Adding an no-op `OpenSearchFunctionExpression` marker to identify this is an expression has no default implement on other runtime (Ex: Prometheus)
 - Update `OpenSearchIndex` in order to provide an OpenSearch specific handler for `eval` operator and its expressions, when OS being used as the storage engine. 

During runtime, all eval expressions, will being passed to `OpenSearchIndex.visitEval( )`, then `OpenSearchEvalOperator` class will pick up the call, by evaluating all `eval` expression as it is, and then handle all occasion of `OpenSearchFunctionExpression` separately, by reading the function name and argument, and execute the appropriate business logic. 

Marker class `OpenSearchFunctionExpression` is being used in this case because the actual implementation require runtime OpenSearch client connectivity, however `core` module is mean to be generic, hence this workaround is being deployed, by tagging it as `OpenSearchFunctionExpression` on `core` and only handle it on the `opensearch` Cradle module .

### Related Issues
Resolves: https://github.com/opensearch-project/sql/issues/3037

### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
 - [x] New functionality has javadoc added.
 - [x] New functionality has a user manual doc added.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/sql/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by andy-k-improving** [comment]
As per the offline discussion, I have separated out the integration related changes into https://github.com/opensearch-project/sql/pull/3244, in order to minimise the diff.

File: docs/user/ppl/functions/geoip.rst [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+---------
+
+Description
+>>>>>>>>>>>
+
+Usage: `geoip(dataSourceName, ipAddress, options)` to lookup location information from given IP addresses via OpenSearch GeoSpatial plugin API.
+
+Argument type: STRING, STRING, STRING
+
+Return type: Tuple
```
Comment (acarbonetto) [inline docs/user/ppl/functions/geoip.rst:21]:
```suggestion
Return type: OBJECT
```

File: docs/user/ppl/functions/geoip.rst [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    | LookupResult                                                        |
+    |-------------------------------------------------------------|
+    | {'city_name': 'Vancouver', 'country_iso_code': 'CA'}        |
+    +-------------------------------------------------------------+
+
+
+Note:
+ - `dataSourceName` must be an established dataSource on OpenSearch GeoSpatial plugin, detail of configuration can be found: https://opensearch.org/docs/latest/ingest-pipelines/processors/ip2geo/
+ - `ip` can be an IPv4 or an IPv6 address
+ - `options` is a comma separated String for user to specify fields to output, the selection of fields subject to dataSourceProvider's schema, the list of geolite2-city dataset provide fields: "country_iso_code", "country_name", "continent_name", "region_iso_code", "region_name", "city_name", "time_zone", "location"
```
Comment (acarbonetto) [inline docs/user/ppl/functions/geoip.rst:37]:
```suggestion
 - `options` is a comma separated String of fields to output: the selection of fields is subject to dataSourceProvider's schema.  For example, the list of fields in the provided `geolite2-city` dataset includes: "country_iso_code", "country_name", "continent_name", "region_iso_code", "region_name", "city_name", "time_zone", "location"
```

Comment (acarbonetto) [reply docs/user/ppl/functions/geoip.rst:37]: Split that sentence, it kind of runs on...

File: core/src/main/java/org/opensearch/sql/expression/ip/GeoIPFunctions.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import org.opensearch.sql.expression.function.FunctionName;
+import org.opensearch.sql.expression.function.FunctionSignature;
+import org.opensearch.sql.expression.function.SerializableFunction;
+
+/**
+ * Utility class to register the method signature for geoip( ) expression, concreted reallocated to
+ * `opensearch` module, as this Ip location require GeoSpatial Plugin runtime support.
+ */
+@UtilityClass
+public class GeoIPFunctions {
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/ip/GeoIPFunctions.java:30]: did we need to create a separate class for GeoIPFunctions when we already have IPFunctions?

File: core/src/main/java/org/opensearch/sql/expression/ip/GeoIPFunctions.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import org.opensearch.sql.expression.function.BuiltinFunctionRepository;
+import org.opensearch.sql.expression.function.DefaultFunctionResolver;
+import org.opensearch.sql.expression.function.FunctionBuilder;
+import org.opensearch.sql.expression.function.FunctionName;
+import org.opensearch.sql.expression.function.FunctionSignature;
+import org.opensearch.sql.expression.function.SerializableFunction;
+
+/**
+ * Utility class to register the method signature for geoip( ) expression, concreted reallocated to
+ * `opensearch` module, as this Ip location require GeoSpatial Plugin runtime support.
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/ip/GeoIPFunctions.java:27]: what do you mean by `concreted reallocated to 'opensearch' module`?

File: docs/user/ppl/functions/geoip.rst [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+   :local:
+   :depth: 1
+
+GEOIP
+---------
+
+Description
+>>>>>>>>>>>
+
+Usage: `geoip(dataSourceName, ipAddress, options)` to lookup location information from given IP addresses via OpenSearch GeoSpatial plugin API.
```
Comment (acarbonetto) [inline docs/user/ppl/functions/geoip.rst:17]:
```suggestion
Usage: `geoip(dataSourceName, ipAddress[, options])` to lookup location information from given IP addresses via OpenSearch GeoSpatial plugin API.
```

Comment (acarbonetto) [reply docs/user/ppl/functions/geoip.rst:17]: It seems like the last argument is optional.

File: core/src/main/java/org/opensearch/sql/expression/ip/OpenSearchFunctionExpression.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import org.opensearch.sql.data.model.ExprValue;
+import org.opensearch.sql.data.type.ExprType;
+import org.opensearch.sql.expression.Expression;
+import org.opensearch.sql.expression.FunctionExpression;
+import org.opensearch.sql.expression.env.Environment;
+import org.opensearch.sql.expression.function.FunctionName;
+
+/**
+ * Marker class to identify functions only compatible with OpenSearch storage engine. Any attempt to
+ * invoke the method different from OpenSearch will result in UnsupportedOperationException.
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/ip/OpenSearchFunctionExpression.java:20]: how does that work?

File: core/src/main/java/org/opensearch/sql/expression/ip/OpenSearchFunctionExpression.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  @Override
+  public ExprValue valueOf() {
+    return null;
+  }
+
+  @Override
+  public ExprValue valueOf(Environment<Expression, ExprValue> valueEnv) {
+    throw new UnsupportedOperationException(
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/ip/OpenSearchFunctionExpression.java:39]: you'll need to add an IT test to show how this works

File: integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java [hunk 8]
```diff
@@ -8,23 +8,48 @@
...
 
 /** IP enrichment PPL request with OpenSearch Geo-sptial plugin */
 public class GeoIpFunctionsIT extends PPLIntegTestCase {
 
   private static boolean initialized = false;
 
+  private static Map<String, Object> MANIFEST_LOCATION =
+          Map.of(
+                  "endpoint",
+                  "https://raw.githubusercontent.com/opensearch-project/geospatial/main/src/test/resources/ip2geo/server/city/manifest.json");
```
Comment (acarbonetto) [inline integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java]: hope this doesn't change... 😬

File: integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java [hunk 9]
```diff
@@ -45,4 +70,86 @@ public void testGeoPluginInstallation() throws IOException {
...
     Assert.assertTrue(getResponseBody(response, true).contains(PLUGIN_NAME));
   }
+
+  @SneakyThrows
+  @Test
+  public void testGeoIpEnrichment() {
+    loadIndex(Index.GEOIP);
+    // Create a new dataSource
+    createDatasource(DATASOURCE_NAME, MANIFEST_LOCATION);
+    waitForDatasourceToBeAvailable(DATASOURCE_NAME, Duration.ofSeconds(10));
```
Comment (acarbonetto) [inline integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java:79]: move `createDatasource` and `waitForDatasourceToBeAvailable` into a `BeforeEach` or `BeforeAll` method so that it doesn't make this test timeout.  

If we add more tests, we shouldn't need to load the datasource each time.

File: integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java [hunk 10]
```diff
@@ -45,4 +70,86 @@ public void testGeoPluginInstallation() throws IOException {
...
+  /**
+   * Helper method to send a PUT request to create a dummy dataSource with provided endpoint for
+   * integration test.
+   *
+   * @param name Name of the dataSource
+   * @param properties Request payload in Json format
+   * @return Response for the create dataSource request.
+   * @throws IOException In case of network failure
+   */
+  private Response createDatasource(final String name, Map<String, Object> properties)
```
Comment (acarbonetto) [inline integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java:103]: Just use the constants...? 
```suggestion
  private Response createDatasource()
```

File: integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java [hunk 11]
```diff
@@ -45,4 +70,86 @@ public void testGeoPluginInstallation() throws IOException {
...
+          throws IOException {
+    XContentBuilder builder = XContentFactory.jsonBuilder().startObject();
+    for (Map.Entry<String, Object> config : properties.entrySet()) {
+      builder.field(config.getKey(), config.getValue());
+    }
+    builder.endObject();
+    Request request = new Request("PUT", GEO_SPATIAL_DATASOURCE_PATH + name);
```
Comment (acarbonetto) [inline integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java:110]:
```suggestion
    Request request = new Request("PUT", GEO_SPATIAL_DATASOURCE_PATH + DATASOURCE_NAME);
```

File: integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java [hunk 12]
```diff
@@ -45,4 +70,86 @@ public void testGeoPluginInstallation() throws IOException {
...
+   *
+   * @param name Name of the dataSource
+   * @param properties Request payload in Json format
+   * @return Response for the create dataSource request.
+   * @throws IOException In case of network failure
+   */
+  private Response createDatasource(final String name, Map<String, Object> properties)
+          throws IOException {
+    XContentBuilder builder = XContentFactory.jsonBuilder().startObject();
+    for (Map.Entry<String, Object> config : properties.entrySet()) {
```
Comment (acarbonetto) [inline integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java:106]:
```suggestion
    for (Map.Entry<String, Object> config : MANIFEST_LOCATION.entrySet()) {
```

File: integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java (hunk 9 above)
Comment (acarbonetto) [inline integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java:79]: Make `10` seconds a constant at the top of the test file.

File: integ-test/src/test/java/org/opensearch/sql/security/CrossClusterSearchIT.java [hunk 13]
```diff
@@ -41,6 +42,7 @@ public class CrossClusterSearchIT extends PPLIntegTestCase {
 
   private static final String TEST_INDEX_BANK_REMOTE = REMOTE_CLUSTER + ":" + TEST_INDEX_BANK;
   private static final String TEST_INDEX_DOG_REMOTE = REMOTE_CLUSTER + ":" + TEST_INDEX_DOG;
+  private static final String TEST_INDEX_GEOIP_REMOTE = REMOTE_CLUSTER + ":" + TEST_INDEX_GEOIP;
```
Comment (acarbonetto) [inline integ-test/src/test/java/org/opensearch/sql/security/CrossClusterSearchIT.java:45]: did you want to add a cross-cluster test here?

File: opensearch/src/main/java/org/opensearch/sql/opensearch/planner/physical/OpenSearchEvalOperator.java [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+  public OpenSearchEvalOperator(
+      PhysicalPlan input,
+      List<Pair<ReferenceExpression, Expression>> expressionList,
+      NodeClient nodeClient) {
+    super(input, expressionList);
+    this.nodeClient = nodeClient;
+  }
+
+  @Override
+  public ExprValue next() {
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/planner/physical/OpenSearchEvalOperator.java:54]: I don't see how this is different from EvalOperator.next()

File: opensearch/src/main/java/org/opensearch/sql/opensearch/planner/physical/OpenSearchEvalOperator.java [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+   * @return The mapping of reference and {@link ExprValue} for each expression.
+   */
+  private Map<String, ExprValue> eval(Environment<Expression, ExprValue> env) {
+    Map<String, ExprValue> evalResultMap = new LinkedHashMap<>();
+    for (Pair<ReferenceExpression, Expression> pair : this.getExpressionList()) {
+      ReferenceExpression var = pair.getKey();
+      Expression valueExpr = pair.getValue();
+      ExprValue value;
+      if (valueExpr instanceof OpenSearchFunctionExpression openSearchFuncExpression) {
+        if ("geoip".equals(openSearchFuncExpression.getFunctionName().getFunctionName())) {
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/planner/physical/OpenSearchEvalOperator.java:88]: can we not fetchIpEnrichment within the geoip function itself?

File: integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java (hunk 9 above)
Comment (andy-k-improving) [reply integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java:79]: Done.

File: docs/user/ppl/functions/geoip.rst [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+Geo IP Address Functions
+====================
+
+.. rubric:: Table of contents
+
+.. contents::
+   :local:
+   :depth: 1
+
+GEOIP
```
[the remark this answers is not in the document]
Comment (andy-k-improving) [reply docs/user/ppl/functions/geoip.rst:11]: Not really, it makes more sense to merge both .rst.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/planner/physical/OpenSearchEvalOperator.java (hunk 14 above)
Comment (andy-k-improving) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/planner/physical/OpenSearchEvalOperator.java]: Removed

File: opensearch/src/main/java/org/opensearch/sql/opensearch/planner/physical/OpenSearchEvalOperator.java (hunk 15 above)
Comment (andy-k-improving) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/planner/physical/OpenSearchEvalOperator.java]: I don't think that is the feasible, as per the current design `geoip` function located in the `core` Gradle sub-module which is not mean to be storage engine specific implementation. 
Hence for the alternative, I created `OpenSearchEvalProcessor` to serve a central registry for all `eval` OpenSearch specific operation. 
WDYT?

File: docs/user/ppl/functions/geoip.rst (hunk 16 above)
[the remark this answers is not in the document]
Comment (andy-k-improving) [reply docs/user/ppl/functions/geoip.rst]: Actually my initial idea was that, since it's not possible to create the test  dataSource (Not dataSet) in programmatically manner during the docTest run (Unlike integration-test I can setup the dataSource on Junit).
As the result, I will need to exclude this `geoip.rst` from the DocTest, and if this got merged into `ip.rst`, the original DoctTest validation from `ip.rst` will also be skipped.

File: docs/category.json [hunk 17]
```diff
@@ -33,7 +33,6 @@
     "user/ppl/functions/condition.rst",
     "user/ppl/functions/datetime.rst",
     "user/ppl/functions/expressions.rst",
-    "user/ppl/functions/ip.rst",
```
Comment (normanj-bitquill) [inline docs/category.json:36]: Why is this removed?

File: docs/user/ppl/functions/geoip.rst (hunk 16 above)
[the remark this answers is not in the document]
Comment (acarbonetto) [reply docs/user/ppl/functions/geoip.rst]: If the test cannot be run as part of a doctest run, then exclude the `os>`.  Then the example won't be compiled and run as part of doctest. If you can add a comment in the code "why" it's being excluded, that would be nice.

File: core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java [hunk 18]
```diff
@@ -130,33 +130,41 @@ private static FunctionResolver score(BuiltinFunctionName score) {
   public static class OpenSearchFunction extends FunctionExpression {
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java:130]: Alternatively, can you create a similar class (maybe `OpenSearchExecutableFunction` or `OpenSearchExecutable`) as opposed to re-using this class.  This could represent the functionexpressions that are 'executable' within an `eval` statement (only used for your one method... for now). 

Then under `OpenSearchEvalOperator.eval`, you'd look for `OpenSearchExecutableFunction` and call `process`... and it wouldn't affect all the relevance methods.  

Note: to get it to work in `where` and `having`, extend `OpenSearchFunction`.

Comment (acarbonetto) [reply core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java:130]: Potentially, we should rename `OpenSearchFunction` to something like `OpenSearchBooleanFunction`... but that's probably out of scope.

Comment (acarbonetto) [reply core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java:130]: That being said, making relevance functions work inside `eval` would be a decent feature...

File: core/src/main/java/org/opensearch/sql/expression/ip/IPFunctions.java [hunk 19]
```diff
@@ -57,4 +67,34 @@ private ExprValue exprCidrMatch(ExprValue addressExprValue, ExprValue rangeExprV
...
+
+  /**
+   * Util method to generate probe implementation with given list of argument types, with marker
+   * class `OpenSearchFunction` to annotate this is an OpenSearch specific expression.
+   *
+   * @param returnType return type.
+   * @return Binary Function Implementation.
+   */
+  public static SerializableFunction<FunctionName, Pair<FunctionSignature, FunctionBuilder>>
+      openSearchImpl(ExprType returnType, List<ExprType> args) {
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/ip/IPFunctions.java:91]: move this to `OpenSearchFunctions.java`

File: docs/category.json (hunk 17 above)
Comment (acarbonetto) [reply docs/category.json:36]: if your new test cannot be run within doctest, remove `os>` from the document.  

If you can include a TODO for which this cannot be run... please add the comment.

File: docs/user/ppl/functions/geoip.rst (hunk 16 above)
[the remark this answers is not in the document]
Comment (andy-k-improving) [reply docs/user/ppl/functions/geoip.rst]: Added comment with `os>` removed.

File: docs/category.json (hunk 17 above)
Comment (andy-k-improving) [reply docs/category.json:36]: Done.

File: core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java (hunk 18 above)
Comment (andy-k-improving) [reply core/src/main/java/org/opensearch/sql/expression/function/OpenSearchFunctions.java]: Done.

**Review by normanj-bitquill — APPROVED** [review]
Looks good to me.

File: docs/user/ppl/functions/ip.rst [hunk 20]
```diff
@@ -36,3 +36,34 @@ Note:
...
+    | LookupResult                                                |
+    |-------------------------------------------------------------|
+    | {'city_name': 'Vancouver', 'country_iso_code': 'CA'}        |
+    +-------------------------------------------------------------+
+
+
+Note:
+ - `dataSourceName` must be an established dataSource on OpenSearch GeoSpatial plugin, detail of configuration can be found: https://opensearch.org/docs/latest/ingest-pipelines/processors/ip2geo/
+ - `ip` can be an IPv4 or an IPv6 address
+ - `options` is a comma separated String of fields to output: the selection of fields is subject to dataSourceProvider's schema.  For example, the list of fields in the provided `geolite2-city` dataset includes: "country_iso_code", "country_name", "continent_name", "region_iso_code", "region_name", "city_name", "time_zone", "location"
```
Comment (acarbonetto) [inline docs/user/ppl/functions/ip.rst:68]:
```suggestion
 - `options` is an optional String of comma separated fields to output: the selection of fields is subject to dataSourceProvider's schema.  For example, the list of fields in the provided `geolite2-city` dataset includes: "country_iso_code", "country_name", "continent_name", "region_iso_code", "region_name", "city_name", "time_zone", "location"
```

File: integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java [hunk 21]
```diff
@@ -45,4 +74,79 @@ public void testGeoPluginInstallation() throws IOException {
     Assert.assertEquals(200, response.getStatusLine().getStatusCode());
     Assert.assertTrue(getResponseBody(response, true).contains(PLUGIN_NAME));
   }
+
+  @SneakyThrows
+  @Test
+  public void testGeoIpEnrichment() {
```
Comment (acarbonetto) [inline integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java:80]: we need a test with `options` defined

Comment (acarbonetto) [reply integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java:80]: can you also test whitespace in the comma-separated list?  like: 
```
"country_iso_code,city_name"
"country_iso_code , city_name"
```

File: opensearch/src/main/java/org/opensearch/sql/opensearch/planner/physical/OpenSearchEvalOperator.java [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+import org.opensearch.sql.expression.Expression;
+import org.opensearch.sql.expression.ReferenceExpression;
+import org.opensearch.sql.expression.env.Environment;
+import org.opensearch.sql.expression.function.OpenSearchFunctions;
+import org.opensearch.sql.planner.physical.EvalOperator;
+import org.opensearch.sql.planner.physical.PhysicalPlan;
+
+/**
+ * OpenSearch version of eval operator, which contains nodeClient, in order to perform OpenSearch
+ * related operation during the eval process.
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/planner/physical/OpenSearchEvalOperator.java:28]:
```suggestion
 * specific operations during the eval process.
```

File: integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java (hunk 21 above)
Comment (andy-k-improving) [reply integ-test/src/test/java/org/opensearch/sql/ppl/GeoIpFunctionsIT.java:80]: Added 2 more use-tests, one with single option, and the other one with multiple options separated in comma and space.

**Review by acarbonetto — APPROVED** [review]
lgtm
